File: campus-exchange/mobile/lib/models/student.dart
class Student {
  final String studentId;
  final String collegeId;
  final String collegeName;
  final String officialEmail;
  final String fullName;
  final String department;
  final String verificationStatus;
  final String accountStatus;
  final String role;
  final String? rollNumber;
  final DateTime? verifiedAt;

  Student({
    required this.studentId,
    required this.collegeId,
    required this.collegeName,
    required this.officialEmail,
    required this.fullName,
    required this.department,
    required this.verificationStatus,
    required this.accountStatus,
    required this.role,
    this.rollNumber,
    this.verifiedAt,
  });

  factory Student.fromJson(Map<String, dynamic> json) {
    return Student(
      studentId: json['studentId'] ?? '',
      collegeId: json['collegeId'] ?? '',
      collegeName: json['collegeName'] ??
          'Avanthi Institute of Engineering and Technology',
      officialEmail: json['officialEmail'] ?? '',
      fullName: json['fullName'] ?? 'Verified Student',
      department: json['department'] ?? 'Engineering',
      verificationStatus: json['verificationStatus'] ?? 'pending',
      accountStatus: json['accountStatus'] ?? 'active',
      role: json['role'] ?? 'student',
      rollNumber: json['rollNumber'],
      verifiedAt: json['verifiedAt'] != null
          ? DateTime.tryParse(json['verifiedAt'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'studentId': studentId,
      'collegeId': collegeId,
      'collegeName': collegeName,
      'officialEmail': officialEmail,
      'fullName': fullName,
      'department': department,
      'verificationStatus': verificationStatus,
      'accountStatus': accountStatus,
      'role': role,
      'rollNumber': rollNumber,
      'verifiedAt': verifiedAt?.toIso8601String(),
    };
  }

  bool get isVerified => verificationStatus == 'verified';

  bool get isAdmin => role == 'admin';

  bool get isModerator => role == 'moderator';

  bool get canAccessAdminPanel => isAdmin || isModerator;
}

File: server/src/models/Student.js
const { VERIFICATION_STATUS, ACCOUNT_STATUS, USER_ROLE } = require('../config/constants');
const mongoose = require('mongoose');

const StudentSchema = new mongoose.Schema(
  {
    studentId: {
      type: String,
      required: true,
      unique: true,
      index: true,
    },
    collegeId: {
      type: String,
      required: true,
      index: true,
      ref: 'College',
    },
    officialEmail: {
      type: String,
      required: true,
      trim: true,
      lowercase: true,
      index: true,
    },
    rollNumber: {
      type: String,
      required: true,
      trim: true,
      uppercase: true,
    },
    fullName: {
      type: String,
      trim: true,
      default: 'Student',
    },
    department: {
      type: String,
      trim: true,
      default: 'General Engineering',
    },
    graduatingYear: {
      type: Number,
      default: 2026,
    },
    verificationStatus: {
      type: String,
      enum: Object.values(VERIFICATION_STATUS),
      default: VERIFICATION_STATUS.PENDING,
      index: true,
    },
    accountStatus: {
      type: String,
      enum: Object.values(ACCOUNT_STATUS),
      default: ACCOUNT_STATUS.ACTIVE,
      index: true,
    },
    role: {
      type: String,
      enum: Object.values(USER_ROLE),
      default: USER_ROLE.STUDENT,
      index: true,
    },
    warningCount: {
      type: Number,
      default: 0,
      min: 0,
    },
    verifiedAt: {
      type: Date,
    },
  },
  {
    timestamps: true,
  }
);

// Compound indexes: rollNumber and officialEmail must be unique per college
StudentSchema.index({ collegeId: 1, rollNumber: 1 }, { unique: true });
StudentSchema.index({ collegeId: 1, officialEmail: 1 }, { unique: true });

module.exports = mongoose.model('Student', StudentSchema);

